[refactor] Refactor AudioInput to use useWaveformController

sfc-gh-nbellante opened the pull request:
## Describe your changes

**TLDR: This PR uses the foundation provided by** #12727 **and** #12728 **to simplify the amount of code needed to make an audio input widget (and by extension the chat input version). I also added some extra testing while noticing some regressions in the process that didn't get caught specifically around the timer behavior.**

This pull request primarily improves the reliability and maintainability of the audio input widget, focusing on test robustness, UI styling, and code cleanup. The most significant changes include making end-to-end tests for the audio input timer display more robust to timing variations, introducing a new unit test for the timer display in the frontend, and enhancing the styling of waveform components to better handle overlapping elements. Additionally, a legacy audio conversion utility has been removed as it is no longer in use.

**Test robustness and coverage improvements:**

- Refactored `test_audio_input_timer_display` in `st_audio_input_test.py` to be more robust against timing variations by adding buffer ranges, regex-based assertions, and a new `wait_for_timer_at_least` helper, reducing flakiness in CI and local runs. [[1]](diffhunk://#diff-3c2dea8beeab487a4a39239f1224185cf66a32c39a79a62cf09637a8a044a980L615-R636) [[2]](diffhunk://#diff-3c2dea8beeab487a4a39239f1224185cf66a32c39a79a62cf09637a8a044a980L639-R665)
- Added a new frontend unit test `AudioInput.test.tsx` to verify that the timer display updates correctly during recording progress events, increasing test coverage and reliability.

**UI and styling enhancements:**

- Improved `StyledWaveSurferDiv` in `styled-components.ts` to stack multiple WaveSurfer instances with proper positioning and flex alignment, ensuring correct visual overlap of recording and playback waveforms.
- Added `overflow: "hidden"` to `StyledWaveformContainerDiv` and `flex: 1` to `StyledPlaceholderContainerDiv` to improve layout consistency and prevent visual overflow issues. [[1]](diffhunk://#diff-4b7b63c4a64aef0de9b7dbd7b0d943c4ae9aa183a51715b633f07e69a6a26ab8R37) [[2]](diffhunk://#diff-4b7b63c4a64aef0de9b7dbd7b0d943c4ae9aa183a51715b633f07e69a6a26ab8R98)

**Code cleanup:**

- Removed the unused `convertAudioToWav.ts` utility, which contained legacy audio conversion logic, as it is no longer referenced in the codebase.

## Testing Plan

- E2E Tests were improved and ensured no regressions in this refactor.

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

sfc-gh-nbellante committed:
Refactor AudioInput to use useWaveformController

This completes the refactoring by updating AudioInput to use the new
headless waveformController hook:

- Replace direct WaveSurfer manipulation with controller API
- Simplify component logic with declarative event handlers
- Improve memory management and cleanup
- Update widget tests to work with new architecture
- Adjust E2E test for more deterministic timer behavior

No UI changes - this is a pure refactor for better maintainability.

sfc-gh-nbellante commented:
* **#12729** <a href="https://app.graphite.dev/github/pr/streamlit/streamlit/12729?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/streamlit/streamlit/12729?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#12728** <a href="https://app.graphite.dev/github/pr/streamlit/streamlit/12728?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#12727** <a href="https://app.graphite.dev/github/pr/streamlit/streamlit/12727?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#12726** <a href="https://app.graphite.dev/github/pr/streamlit/streamlit/12726?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `develop`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

kmcgrady commented on frontend/lib/src/components/widgets/AudioInput/AudioInput.tsx:
```diff
@@ -86,12 +74,13 @@ const AudioInput: React.FC<Props> = ({
-  const waveSurferRef = useRef<HTMLDivElement | null>(null)
-  // Track current blob URL for cleanup
-  const currentBlobUrlRef = useRef<string | null>(null)
+  const containerRef = useRef<HTMLDivElement>(null)
```
Shouldn't this be `useRef<HTMLDivElement | null>`?

sfc-gh-nbellante replied:
Actually, the current type is already takes that into account! When you write:

```typescript
const containerRef = useRef<HTMLDivElement>(null)
```

React's type definitions automatically give you `RefObject<HTMLDivElement>` which has `current: HTMLDivElement | null`. The null is handled by React's function overload resolution - when you pass `null` as the initial value, it uses the overload that returns `RefObject<T>` (not `MutableRefObject<T>`).

sfc-gh-nbellante replied:
I also did not know this until Claude pointed it out to me which is why the previous ones had `| null`. 😅

kmcgrady commented on frontend/lib/src/components/widgets/AudioInput/AudioInput.test.tsx:
```diff
@@ -0,0 +1,139 @@
+    })
+
+    await waitFor(() => expect(timer).toHaveTextContent("00:03"))
+  })
```
Tests seem like. Is it because most are done in the React hook, and it's hard to test any other aspects of AudioInput?

sfc-gh-nbellante replied:
It was because it was tested in the hook, but JS tests are cheap so I don't mind adding tests here too.

sfc-gh-nbellante committed:
Add comprehensive unit tests for AudioInput component

This commit addresses code review feedback about insufficient test
coverage in AudioInput.test.tsx. Previously only 1 test existed,
which was too light given the component's responsibilities.

Added 13 new tests covering:
- Timer display updates (7 tests): recording progress, playback timing,
  duration updates, pause state, and cancel behavior
- Error handling (2 tests): onError events and permission denied states
- File upload flow (1 test): blob URL creation on approve
- Cleanup behavior (2 tests): animation frame cleanup on unmount and
  playback stop
- Button interactions (2 tests): record button triggers start, stop
  button triggers stop+approve

While most recording/playback logic is now in useWaveformController
(tested separately), AudioInput still has significant glue code that
needs testing: file upload with abort handling, memory management
(blob URLs, animation frames), integration with WidgetStateManager,
error state management, and event handler wiring.

Total test count: 14 tests, all passing.

sfc-gh-nbellante committed:
Add comprehensive unit tests for AudioInput component

Expanded test coverage from 1 to 18 tests to address code review feedback.
Tests are organized by user journey and cover:

- Recording flow (start, stop, cancel, replace, timer updates)
- Playback flow (play, pause, resume, timer updates)
- Upload & abort handling (unmount, new upload, errors)
- Memory management (blob URL cleanup, animation frame cleanup)
- Form integration (clear on submit)
- Error handling (permission denied, generic errors)
- Widget state management

All tests follow React Testing Library best practices with proper
async handling and user-centric queries.
